src/DNS_Mgr.cc

```diff
@@ -992,23 +992,54 @@ void DNS_Mgr::Resolve()
 	{
 	int nfds = 0;
 	struct timeval *tvp, tv;
-	fd_set read_fds, write_fds;
+	struct pollfd pollfds[ARES_GETSOCK_MAXNUM];
+	ares_socket_t socks[ARES_GETSOCK_MAXNUM];
 
 	tv.tv_sec = DNS_TIMEOUT;
 	tv.tv_usec = 0;
 
 	for ( int i = 0; i < MAX_PENDING_REQUESTS; i++ )
 		{
-		FD_ZERO(&read_fds);
-		FD_ZERO(&write_fds);
-		nfds = ares_fds(channel, &read_fds, &write_fds);
+		int nfds = 0;
+		int bitmap = ares_getsock(channel, socks, ARES_GETSOCK_MAXNUM);
+
+		for ( int i = 0; i < ARES_GETSOCK_MAXNUM; i++ )
+			{
+			bool rd = ARES_GETSOCK_READABLE(bitmap, i);
+			bool wr = ARES_GETSOCK_WRITABLE(bitmap, i);
+			if ( rd || wr )
+				{
+				pollfds[nfds].fd = socks[i];
+				pollfds[nfds].events = rd ? POLLIN : 0;
+				pollfds[nfds].events |= wr ? POLLOUT : 0;
+				++nfds;
+				}
+			}
+
+		// Do we have any sockets that are read or writable?
 		if ( nfds == 0 )
 			break;
 
+		// poll() timeout is in milliseconds.
 		tvp = ares_timeout(channel, &tv, &tv);
-		int res = select(nfds, &read_fds, &write_fds, NULL, tvp);
-		if ( res >= 0 )
-			ares_process(channel, &read_fds, &write_fds);
+		int timeout_ms = tvp->tv_sec * 1000 + tvp->tv_usec / 1000;
+
+		int res = poll(pollfds, nfds, timeout_ms);
```

Comment:
Tim asked why we're using poll() directly instead of going through kqueue. It had been using `select()` directly before, so that's not much different. It's the code path used for synchronous lookups and during shutdown, so it was probably difficult to plumb together with the async IO loop.